bolt-design-system/bolt
# Set up Docker deployment and Express Server

## Jira

- BDS-693
- BDS-586

## Summary

Replacing static site hosting with an Express server and a Docker deployment and wiring up the Bolt API!

## Details

Travis CI builds a Docker image that runs the install, build, and test and then pushes it up [to Docker Hub](https://cloud.docker.com/u/boltdesignsystem/repository/docker/boltdesignsystem/bolt), then our deploy script goes into the `./deploys/` folder to run the now.sh deploy that pulls that Docker image down to run and serve. This is done to get around issues that now's deploy script has with upload all the correct files due to it's incorrect parsing of ignore files. Additionally, all build dependencies like php, node, and yarn version are pre-built into the `bolt-docker` image (details on [Docker Hub](https://cloud.docker.com/u/boltdesignsystem/repository/docker/boltdesignsystem/bolt-docker) and the [GitHub repo](https://github.com/bolt-design-system/bolt-docker)).

Additionally, the route that was available at `/api` when running the local server is now hooked up to the hosted site! It's using the same routing so all should be available locally and on the deployed server.

I've also added the now.sh unique deploy urls as a commit status check to each commit so you can view the unique url each commit generated.

## How to test

All of this relates to CI work so checking Travis logs and its deployed site should be enough. However, one could build and run the Docker container using these commands:

```bash
bash ./scripts/docker-build.sh
bash ./scripts/docker-run.sh
```

To test the api, send a GET request to `/api` and you'll get back a simple response saying you've hit the Bolt API. The only other route that was being used locally, `/api/render-twig` is also available, but ensuring that it works on server as it does with a local server is outside the scope of this PR. 

## Remaining Work

- [x] Ensure redirects are in place:

https://github.com/bolt-design-system/bolt/blob/aebfb651e250e0fab1610662622b3b56b773ad33/Xnow.old.json#L8-L37

- [ ] Ensure that old deploys are not using the accounts finite amount of minimum [scaling](https://zeit.co/docs/api/v1/#endpoints/deployments/set-scale-configuration). I have the minimum set to 1 so the site loads quick. I could see setting old deploys that aren't the latest on a branch to a minimum scale of 0 so you could still access them but they didn't take up one of the "Instances" you have (pretty sure you're on [Pro plan with 25 available instances](https://zeit.co/pricing/v1)). Also, it's possible to clean up old deploys with:

```bash
now rm boltdesignsystem --safe --yes # --safe skips removing ones with active alias
```

- [x] Speed up CI. This looks like it's adding more time than I'd like.

- [ ] Consider how to handle NPM deps used in Deploy and Nightwatch scripts. I'm skipping the `yarn install` in the Travis build as the Docker build does that, so when our scripts that deploy and kick off Nightwatch testing start, they don't have. Currently I've set up the `./scripts` folders own `package.json` as a workaround:

https://github.com/bolt-design-system/bolt/blob/7d983c69d98caf2d3fc0623ce22bc8f18a598ebf/.travis.yml#L37-L40
